### PR TITLE
449 - make constants instructions rather that expressions

### LINKED ISSUE #449: Micheline/Michelson converter
Conseil is not processing literal parameters correctly resulting in things of this nature:

```
Unparsable code: {"int":"1"}
Unparsable code: [{"string":"2019-07-02T16:17:38Z"},{"string":"2019-07-02T16:17:08Z"}]
Unparsable code: {"string":"hello"}
```

### PR DESCRIPTION
**Pull request by dorians**
closes #449 

This PR changes the way constant types are parsed (as an instruction, not type).

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/dto/MichelsonInstruction.scala [hunk 1]
```diff
@@ -32,8 +34,17 @@ case class MichelsonSingleInstruction(
 /* Class representing a sequence of Michelson instructions */
 case class MichelsonInstructionSequence(instructions: List[MichelsonInstruction] = List.empty)
     extends MichelsonInstruction {
-  lazy val normalized: MichelsonInstruction = if (instructions.isEmpty) MichelsonEmptyInstruction else this
+  override lazy val normalized: MichelsonInstruction = if (instructions.isEmpty) MichelsonEmptyInstruction else this
 }
 
+/* Class representing an int constant */
+case class MichelsonIntConstant(int: Long) extends MichelsonInstruction
```
Comment (anonymoussprocket) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/dto/MichelsonInstruction.scala]: "Literal" instead of "Constant" would be better here.

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 2]
```diff
@@ -120,25 +98,51 @@ object JsonParser {
...
-        }
+        embeddedElements = args.getOrElse(List.empty).map(toMichelsonElement)
       )
   }
 
   case class JsonInstructionSequence(instructions: List[JsonInstruction]) extends JsonInstruction {
     override def toMichelsonInstruction = MichelsonInstructionSequence(instructions.map(_.toMichelsonInstruction))
   }
 
+  /*
```
Comment (anonymoussprocket) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: There are other types, `address`, `nat`, boolean, to name a few, do they need to be implemented separately?

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/dto/MichelsonInstruction.scala (hunk 1 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/dto/MichelsonInstruction.scala]: Probably yes, but in the docs, they use word "Constant".

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 2 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: I'm not sure how the types are represented because it's not documented so I created those structures only for `int`, `string` and `bytes`. They were the only constructions for which parsing failed.

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/dto/MichelsonInstruction.scala [hunk 3]
```diff
@@ -32,8 +34,17 @@ case class MichelsonSingleInstruction(
 /* Class representing a sequence of Michelson instructions */
 case class MichelsonInstructionSequence(instructions: List[MichelsonInstruction] = List.empty)
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/dto/MichelsonInstruction.scala]: Might not be the best idea to spend time on this... but, since the parser will never call this with an empty list arg, why not changing the constructor to only accept non empty lists? varargs can be used to do that.
Would this avoid the need for a normalized form wholesale?

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 4]
```diff
@@ -38,6 +39,17 @@ object JsonParser {
     def toMichelsonExpression: MichelsonExpression
   }
 
+  type EmbeddedElement = Either[Either[JsonExpression, JsonInstruction], List[JsonInstruction]]
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: Minor: Is there a way to use an ADT for this instead of "nested disjunctions"? Personally I find them hard to parse mentally and to reason with.

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 5]
```diff
@@ -120,25 +98,51 @@ object JsonParser {
...
+  }
+
+  /*
+   * Wrapper for bytes constant
+   *
+   * {"bytes": "0500"}
+   *
+   * */
+  case class JsonBytesConstant(bytes: String) extends JsonInstruction {
+    override def toMichelsonInstruction = MichelsonBytesConstant(bytes)
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: Just to clarify, does this value encode byte-strings? That is, multiple byte-encoded values?
